kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -203,15 +225,27 @@ public Optional<SetTransaction> getSetTxnOpt() {
         return setTxnOpt;
     }
 
-    private Row generateCommitAction() {
+    private Optional<Long> generateInCommitTimestampForFirstCommitAttempt(
+            Engine engine, long currentTimestamp) {
+        if (IN_COMMIT_TIMESTAMPS_ENABLED.fromMetadata(metadata)) {
+            return Optional.of(currentTimestamp);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private CommitInfo generateCommitAction(Engine engine) {
+        long commitAttemptStartTime = System.currentTimeMillis();
         return new CommitInfo(
-                System.currentTimeMillis(), /* timestamp */
+                generateInCommitTimestampForFirstCommitAttempt(
```

Comment:
is the change to look at the last commit and construct the new timestamp coming in next PRs?

Reply:
Yes.